source/module_esolver/esolver_ks_pw.cpp

```diff
@@ -224,6 +221,17 @@ void ESolver_KS_PW<T, Device>::deallocate_hsolver()
     this->phsol = nullptr;
 }
 template <typename T, typename Device>
+void ESolver_KS_PW<T, Device>::allocate_hamilt()
```

Comment:
do not add new functions in ESolver

Reply:
> do not add new functions in ESolver

Because `Hamilt phamilt` is a member of ESolver, this function cannot be avoided to support both `HamiltPW` and `HamiltLIP` in the same `ESolver_KS_PW`, unless I copy the whole `before_scf()`;